### ISSUE TITLE
[sdk] Complier error: No member named `kConstExprCategoryRegistry` in namespace `perfetto::internal`

### ISSUE THREAD

**Issue by sudara** [google/perfetto#164]
I'm having trouble getting the `TRACE_EVENT` macro compiling on MacOS using tag v17.0 or v18.0. Initialization and config setup all seems otherwise happy:

```
/Users/sudara/projects/myproject/Source/Components/myfile.h:24:9: error: no member named 'kConstExprCategoryRegistry' in namespace 'perfetto::internal'
        TRACE_EVENT("component", "myFunction");
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/sudara/projects/myproject/cmake-build-debug/_deps/perfetto-src/sdk/perfetto.h:15995:3: note: expanded from macro 'TRACE_EVENT'
  PERFETTO_INTERNAL_SCOPED_TRACK_EVENT(category, name, ##__VA_ARGS__)
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/sudara/projects/myproject/cmake-build-debug/_deps/perfetto-src/sdk/perfetto.h:15673:27: note: expanded from macro 'PERFETTO_INTERNAL_SCOPED_TRACK_EVENT'
      ~EventFinalizer() { TRACE_EVENT_END(category); }                        \
                          ^~~~~~~~~~~~~~~~~~~~~~~~~
/Users/sudara/projects/myproject/cmake-build-debug/_deps/perfetto-src/sdk/perfetto.h:15989:3: note: expanded from macro 'TRACE_EVENT_END'
  PERFETTO_INTERNAL_TRACK_EVENT(       \
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/sudara/projects/myproject/cmake-build-debug/_deps/perfetto-src/sdk/perfetto.h:15642:9: note: expanded from macro 'PERFETTO_INTERNAL_TRACK_EVENT'
        PERFETTO_GET_CATEGORY_INDEX(category);                                 \
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/sudara/projects/myproject/cmake-build-debug/_deps/perfetto-src/sdk/perfetto.h:15624:47: note: expanded from macro 'PERFETTO_GET_CATEGORY_INDEX'
  ::PERFETTO_TRACK_EVENT_NAMESPACE::internal::kConstExprCategoryRegistry.Find( \
```

I've added the category, as the [example](https://android.googlesource.com/platform/external/perfetto/+/refs/heads/simpleperf-release/examples/sdk/example.cc) does.

```
    #include <perfetto.h>
    PERFETTO_DEFINE_CATEGORIES(
    perfetto::Category("component")
        .SetDescription("Component"));
```

**Comment by chromy** [comment]
Hi Sudara!

Do you declare `PERFETTO_TRACK_EVENT_STATIC_STORAGE();` in a `.cc` file somewhere?
(for example see the second code snipit on https://perfetto.dev/docs/instrumentation/track-events). If not I believe this may be your issue. 😃

**Comment by sudara** [comment]
Thanks for the response @chromy!

Yes, I have `PERFETTO_TRACK_EVENT_STATIC_STORAGE();` in a .cpp.

Initially I didn't realize categories were mandatory, so was I missing `PERFETTO_DEFINE_CATEGORIES` and saw the following compiler error for `perfetto::TrackEvent::Register();`: `no member named 'TrackEvent' in namespace 'perfetto'`

However, I looked at the example and setup an extra header for the categories and .cpp file to call `PERFETTO_TRACK_EVENT_STATIC_STORAGE();` — if i fail to call that at all, I get a linker error instead.

```
Undefined symbols for architecture x86_64:
  "perfetto::DataSource<perfetto::TrackEvent, perfetto::internal::TrackEventDataSourceTraits>::static_state_", referenced from:
      bool perfetto::DataSource<perfetto::TrackEvent, perfetto::internal::TrackEventDataSourceTraits>::Register<>(perfetto::protos::gen::DataSourceDescriptor const&) in libMyProject_SharedCode.a(MyProject.cpp.o)
      perfetto::DataSource<perfetto::TrackEvent, perfetto::internal::TrackEventDataSourceTraits>::DefaultTracePointTraits::GetActiveInstances(perfetto::DataSource<perfetto::TrackEvent, perfetto::internal::TrackEventDataSourceTraits>::DefaultTracePointTraits::TracePointData) in libMyProject_SharedCode.a(MyProject.cpp.o)
      void perfetto::DataSource<perfetto::TrackEvent, perfetto::internal::TrackEventDataSourceTraits>::TraceWithInstances<perfetto::DataSource<perfetto::TrackEvent, perfetto::internal::TrackEventDataSourceTraits>::DefaultTracePointTraits, perfetto::internal::TrackEventDataSource<perfetto::TrackEvent, &(perfetto::internal::kCategoryRegistry)>::Flush()::'lambda'(perfetto::DataSource<perfetto::TrackEvent, perfetto::internal::TrackEventDataSourceTraits>::TraceContext)>(unsigned int, &(perfetto::internal::kCategoryRegistry), perfetto::TrackEvent::TracePointData) in libMyProject_SharedCode.a(MyProject.cpp.o)
      void perfetto::DataSource<perfetto::TrackEvent, perfetto::internal::TrackEventDataSourceTraits>::CreateIncrementalStateImpl<perfetto::internal::TrackEventIncrementalState>(perfetto::internal::DataSourceInstanceThreadLocalState*, perfetto::internal::TrackEventIncrementalState const*) in libMyProject_SharedCode.a(MyProject.cpp.o)
  "perfetto::internal::kCategoryRegistry", referenced from:
      perfetto::internal::TrackEventDataSource<perfetto::TrackEvent, &(perfetto::internal::kCategoryRegistry)>::Register() in libMyProject_SharedCode.a(MyProject.cpp.o)
      perfetto::internal::TrackEventDataSource<perfetto::TrackEvent, &(perfetto::internal::kCategoryRegistry)>::OnSetup(perfetto::DataSourceBase::SetupArgs const&) in libMyProject_SharedCode.a(MyProject.cpp.o)
      perfetto::internal::TrackEventDataSource<perfetto::TrackEvent, &(perfetto::internal::kCategoryRegistry)>::OnStop(perfetto::DataSourceBase::StopArgs const&) in libMyProject_SharedCode.a(MyProject.cpp.o)
  "thread-local wrapper routine for perfetto::DataSource<perfetto::TrackEvent, perfetto::internal::TrackEventDataSourceTraits>::tls_state_", referenced from:
      bool perfetto::DataSource<perfetto::TrackEvent, perfetto::internal::TrackEventDataSourceTraits>::Register<>(perfetto::protos::gen::DataSourceDescriptor const&) in libMyProject_SharedCode.a(MyProject.cpp.o)
      void perfetto::DataSource<perfetto::TrackEvent, perfetto::internal::TrackEventDataSourceTraits>::TraceWithInstances<perfetto::DataSource<perfetto::TrackEvent, perfetto::internal::TrackEventDataSourceTraits>::DefaultTracePointTraits, perfetto::internal::TrackEventDataSource<perfetto::TrackEvent, &(perfetto::internal::kCategoryRegistry)>::Flush()::'lambda'(perfetto::DataSource<perfetto::TrackEvent, perfetto::internal::TrackEventDataSourceTraits>::TraceContext)>(unsigned int, &(perfetto::internal::kCategoryRegistry), perfetto::TrackEvent::TracePointData) in libMyProject_SharedCode.a(MyProject.cpp.o)
ld: symbol(s) not found for architecture x86_64
```

So, I still get `kConstExprCategoryRegistry` compiler error when calling `PERFETTO_TRACK_EVENT_STATIC_STORAGE()` from its own file or from my primary cpp file. 

Thanks so much!

**Comment by chromy** [comment]
Could you share the full text of the relevant files? That would help with debugging :)

**Comment by sudara** [comment]
No problem! Thanks for the eyes:

perfetto.h
```
#pragma once

#define PERFETTO 1
#ifdef PERFETTO
    #include <perfetto.h>
    PERFETTO_DEFINE_CATEGORIES(
    perfetto::Category("component")
        .SetDescription("Component"));
#endif
```

perfetto.cpp
```
#include "perfetto.h"

#ifdef PERFETTO
PERFETTO_TRACK_EVENT_STATIC_STORAGE();
#endif
```

In a top level class in the app, as a private member:
```
#ifdef PERFETTO
    std::unique_ptr<perfetto::TracingSession> tracingSession;
#endif
```

In the constructor of that class:
```
#if PERFETTO
    perfetto::TracingInitArgs args;
    args.backends = perfetto::kInProcessBackend;

    perfetto::Tracing::Initialize (args);
    perfetto::TrackEvent::Register();

    // https://github.com/google/perfetto/blob/master/examples/sdk/example.cc
    perfetto::TraceConfig cfg;
    cfg.add_buffers()->set_size_kb (1024);
    auto* ds_cfg = cfg.add_data_sources()->mutable_config();
    ds_cfg->set_name ("track_event");

    tracingSession = perfetto::Tracing::NewTrace();
    tracingSession->Setup (cfg);
    tracingSession->StartBlocking();
#endif
```

My guess is that the problem is that I'm using the TRACE_EVENT macro in other header files included by my main class, but because the main class hasn't run all the initialization yet, the macro somehow isn't happy?

```
        TRACE_EVENT("component", "testing");
```

Just not sure how to approach resolving...Thanks so much for the help!!

**Comment by chromy** [comment]
I think `perfetto::Tracing::Initialize` needs to happen very early (like in main() ideally) ...but I don't understand how that could cause the compiler issue you describe in the OP. Any thing come to mind @skyostil?

**Comment by sudara** [comment]
I got things working. 

Unfortunately I don't know exactly how. I moved thing to a [singleton approach](https://gist.github.com/sudara/a0793715fe45e832abc70a933ea5abc8) (similar to what was popularly done with [chrome tracing](https://gist.github.com/TheCherno/31f135eea6ee729ab5f26a6908eb3a5e)) and i'm beginning/ending sessions in the constructor and destructor of my "main" class. However, it shouldn't be too functionally different from what was there before...

For context, I'm using a [framework](https://juce.com) which has `main()` out of reach, basically you inherit from a class that boots up all sorts of things for you. My goal is to profile code that I'm including myself (not concern with the framework code for now).

Anyway, thanks for the attention, and sorry I couldn't deliver a nice smoking gun... maybe I just needed a clean build and shuffling around files helped with that.
